databricks_cli/pipelines/api.py

```diff
@@ -0,0 +1,157 @@
+        for lib_object in lib_objects:
+            uri_scheme = urllib.parse.urlsplit(lib_object.path).scheme
+            if lib_object.lib_type == 'jar' and uri_scheme == '':
+                local_lib_objects.append(lib_object)
+            elif lib_object.lib_type == 'jar' and uri_scheme.lower() == 'file':
+                local_lib_objects.append(LibraryObject(lib_object.lib_type, lib_object.path[5:]))
+            else:
+                rest_lib_objects.append(lib_object)
+        return local_lib_objects, rest_lib_objects
+
+    @staticmethod
+    def _get_hashed_path(path):
+        hash_buffer = sha1()
+        try:
+            with open(path, 'rb') as f:
+                while True:
+                    data = f.read(BUFFER_SIZE)
+                    if not data:
+                        break
+                    hash_buffer.update(data)
+        except Exception as e:
+            raise RuntimeError('Error \'{}\' while processing {}'.format(e, path))
+
+        file_hash = hash_buffer.hexdigest()
+        path = 'dbfs:/pipelines/code/{}{}'.format(file_hash, os.path.splitext(path)[1])
+        return path
+
+    def _get_remote_mappings(self, local_lib_objects):
+        return list(map(lambda llo: LibraryObject(llo.lib_type, self._get_hashed_path(llo.path)),
+                        local_lib_objects))
+
+    def _get_files_to_upload(self, local_lib_objects, remote_lib_objects):
+        transformed_remote_lib_objects = map(
+            lambda rlo: LibraryObject(rlo.lib_type, DbfsPath(rlo.path)),
+            remote_lib_objects)
+        return list(filter(lambda lo_tuple: not self.dbfs_client.file_exists(lo_tuple[1].path),
+                           zip(local_lib_objects, transformed_remote_lib_objects)))
+
+    #
+    # Only required until the deploy/delete APIs requires the credentials in the body as well
+    # as the header.Once the API requirement is relaxed, this function can be stripped out and
+    # includes for this function removed.
+    #
+    @staticmethod
+    def _get_credentials_for_request():
+        profile = get_profile_from_context()
+        if profile:
+            config = ProfileConfigProvider.get_config(profile)
+        else:
+            config = get_config()
+        if not config or not config.is_valid:
+            raise InvalidConfigurationError.for_profile(profile)
+
+        if config.is_valid_with_token:
+            return {'token': config.token}
+        else:
+            return {'user': config.username, 'password': config.password}
+
+    def deploy(self, spec, headers=None):
+        lib_objects = LibraryObject.convert_from_libraries(spec.get('libraries', []))
+        local_lib_objects, rest_lib_objects = \
+            self._partition_libraries_and_extract_local_paths(lib_objects)
+        remote_lib_objects = self._get_remote_mappings(local_lib_objects)
+        upload_files = self._get_files_to_upload(local_lib_objects, remote_lib_objects)
```

Comment:
Might be hard to fold together because the return values are needed further down in the computation. 
I've removed _get_remote_mappings because it was a simple map on a list and documented all other functions.